Initial support for EFR32BG22

lpawelcz opened the pull request:
This PR supersedes PR #35361 and is a minimal version thereof. Its main purpose is to introduce basic support for the BRD4184B board using the EFR32BG22 SoC.

The PR consists of:
* Pinctrl driver
* Modifications to the UART driver
* Sleep Timer (Real-Time Counter) driver and tests support
* EFR32BG22 SoC support
* EFR32BG22-SLTB010A board support

Initial support was verified on hardware by running the following samples:
* hello_world
* blinky
* counter/alarm
* subsys/shell/shell_module

This PR should be merged only after https://github.com/zephyrproject-rtos/hal_silabs/pull/19 is merged and `hal_silabs` revision in [west.yml](https://github.com/zephyrproject-rtos/zephyr/blob/1a9c7ecea15ccf6c9cc7393209120aaae847e82e/west.yml#L122) is bumped to the newest one.

fkokosinski committed:
drivers: serial: uart_gecko: Make driver dependent on pinctrl

This commit adds a series of driver-related changes to
Gecko pinctrl.

Co-authored-by: Mateusz Sierszulski <msierszulski@antmicro.com>
Signed-off-by: Filip Kokosinski <fkokosinski@antmicro.com>

fkokosinski committed:
drivers: pinctrl: Add Silabs Gecko pin controller

This commit adds initial support for gecko pinctrl driver

Co-authored-by: Mateusz Sierszulski <msierszulski@antmicro.com>
Signed-off-by: Filip Kokosinski <fkokosinski@antmicro.com>

fkokosinski committed:
soc: silabs_exx32: Add support for SiLabs EFR32BG22 SoC

This commit adds support for Silicon Labs EFR32BG22 SoC.

Co-authored-by: Mateusz Sierszulski <msierszulski@antmicro.com>
Signed-off-by: Filip Kokosinski <fkokosinski@antmicro.com>

fkokosinski committed:
boards: Add support for Silicon Labs EFR32BG22-SLTB010A board

This commit adds support for Silicon Labs efr32bg22_sltb010a boardo

Co-authored-by: Mateusz Sierszulski <msierszulski@antmicro.com>
Co-authored-by: Pawel Czarnecki <pczarnecki@antmicro.com>
Signed-off-by: Filip Kokosinski <fkokosinski@antmicro.com>
Signed-off-by: Paweł Czarnecki <pczarnecki@antmicro.com>

msierszulski committed:
drivers: hwinfo: Fix Gecko hwinfo driver building

do not build hwinfo driver for this board - it does not have support for it

Signed-off-by: Mateusz Sierszulski <msierszulski@antmicro.com>

msierszulski committed:
drivers: counter: Add counter_gecko_stimer driver

This commit adds initial support for Silabs Real-Time counter

Co-authored-by: Filip Kokosinski <fkokosinski@antmicro.com>
Signed-off-by: Mateusz Sierszulski <msierszulski@antmicro.com>

msierszulski committed:
samples: drivers: counter: support alarm sample in counter_gecko_stimer

This commit adds sample support for the counter_gecko_stimer
driver.

Signed-off-by: Mateusz Sierszulski <msierszulski@antmicro.com>

msierszulski committed:
tests: drivers: counter: counter_basic_api support in counter_gecko_stimer

This commit adds test support for the counter_gecko_stimer
driver.

Signed-off-by: Mateusz Sierszulski <msierszulski@antmicro.com>

str4t0m commented:
@sateeshkotapati ping FYI

lpawelcz committed:
west.yml: hal_silabs: update hal version

This bumps the hal_silabs in order to add initial support for
EFR32BG22 SoCs

Signed-off-by: Paweł Czarnecki <pczarnecki@antmicro.com>

gmarull commented on drivers/pinctrl/pinctrl_gecko.c:
```diff
@@ -0,0 +1,93 @@
+			rxpin.pin = GECKO_GET_PIN(pins[i]);
+			rxpin.mode = gpioModeInput;
+			rxpin.out = 1;
+			soc_gpio_configure(&rxpin);
```
nit: missing header for this API?

lpawelcz replied:
Included `soc_gpio.h`

gmarull commented on dts/bindings/pinctrl/silabs,gecko-pinctrl.yaml:
```diff
@@ -0,0 +1,80 @@
+    description: |
+      Silabs pin controller pin configuration group.
+    properties:
+      psels:
```
nit: `psel` is a terminology found in Nordic SoCs, does Silabs use the same?

lpawelcz replied:
Yes, SiLabs pinctrl is very similar to the one in Nordic SoCs.

gmarull commented on dts/bindings/serial/silabs,gecko-usart.yaml:
```diff
@@ -21,20 +21,16 @@ properties:
 
     location-rx:
       type: array
-      required: true
       description: RX pin configuration defined as <location port pin>
```
note: these properties should actually be flagged as deprecated and removed when pinmux (and any legacy pin config methods) are deleted from tree.

lpawelcz replied:
Marked location properties as deprecated.

gmarull commented on drivers/counter/counter_gecko_stimer.c:
```diff
@@ -0,0 +1,292 @@
+
+#define DT_DRV_COMPAT silabs_gecko_stimer
+
+#include <stddef.h>
+#include <string.h>
+#include <errno.h>
+#include <zephyr/kernel.h>
+#include <zephyr/device.h>
+#include <soc.h>
+#include <em_cmu.h>
+
+#include "sl_sleeptimer.h"
+#include "sl_atomic.h"
+
+#include <zephyr/drivers/counter.h>
+
+#include <zephyr/logging/log.h>
+LOG_MODULE_REGISTER(counter_gecko, CONFIG_COUNTER_LOG_LEVEL);
```
```c
#include <stddef.h>
#include <string.h>
#include <errno.h>

#include <zephyr/kernel.h>
#include <zephyr/device.h>
#include <zephyr/drivers/counter.h>
#include <zephyr/logging/log.h>
#include <soc.h> /* QUESTION: is it really needed? */

#include <em_cmu.h>
#include <sl_sleeptimer.h>
#include <sl_atomic.h>

LOG_MODULE_REGISTER(counter_gecko, CONFIG_COUNTER_LOG_LEVEL);
```

lpawelcz replied:
Fixed includes formatting.

gmarull commented on drivers/counter/counter_gecko_stimer.c:
```diff
@@ -0,0 +1,292 @@
+{
+	ARG_UNUSED(dev);
+
+	return 0;
```
`-ENOTSUP`?

lpawelcz replied:
Implemented proper `counter_stop` and `counter_start` functions.

gmarull commented on dts/arm/silabs/efr32bg22-pinctrl.dtsi:
```diff
@@ -0,0 +1,29 @@
+	usart0_default: usart0_default {
+		group1 {
+			/* configure PE.11 as UART_RX and PE.10 as UART_TX */
+			psels = <GECKO_PSEL(UART_TX, GECKO_PORT_E, GECKO_PIN(10))>,
+				<GECKO_PSEL(UART_RX, GECKO_PORT_E, GECKO_PIN(11))>,
+				<GECKO_LOC(UART_LOC, GECKO_LOCATION(0))>;
```
suggestion: to make code less verbose, you can concatenate (`GECKO_ ##`) in GECKO_PSEL and have something like:

`<GECKO_PSEL(UART_TX, PORT_E, PIN(10))>`

lpawelcz replied:
I see your point and I fully agree but this would require changes across multiple SiLabs SoCs. I think it would be best to do that in a separate PR.

lpawelcz replied:
I'm sorry, you are right, I was looking at `GECKO_LOCATION` for some reason - this macro is used in all other SiLabs SoCs while `GECKO_PSEL` is used only in EFR32BG22. I added the concatenations to the macro definition as you requested, edited relevant part of bindings and modified all usages to take the advantage of shorter args.

gmarull commented on dts/arm/silabs/efr32bg22.dtsi:
```diff
@@ -0,0 +1,178 @@
+#include <dt-bindings/gpio/gpio.h>
+#include <dt-bindings/i2c/i2c.h>
+#include <dt-bindings/pinctrl/gecko-pinctrl.h>
+#include "gpio_gecko.h"
```
?

lpawelcz replied:
Fixed include

gmarull commented on dts/arm/silabs/efr32bg22.dtsi:
```diff
@@ -0,0 +1,178 @@
+			status = "disabled";
+		};
+
+		usart1: usart@40060000 { /* USART1 */
+			compatible = "silabs,gecko-usart";
+			reg = <0x40060000 0x400>;
+			interrupts = <15 0>, <16 0>;
+			interrupt-names = "rx", "tx";
+			peripheral-id = <1>;
+			status = "disabled";
+		};
+
+		leuart0: leuart@4004a000 { /* LEUART0 */
```
redundant comments

lpawelcz replied:
Removed redundant comments

gmarull commented on dts/arm/silabs/efr32bg22.dtsi:
```diff
@@ -0,0 +1,178 @@
+			compatible = "silabs,gecko-leuart";
+			reg = <0x4004a000 0x400>;
+			interrupts = <22 0>;
+			peripheral-id = <0>;
```
What is this "peripheral-id" used for?

lpawelcz replied:
It's used for indexing if there are multiple peripherals of given type. For example in GPIO driver it is used for getting `gpio_index`. In LEUART it is used to distinguish between `cmuClock_LEUART0` and `cmuClock_LEUART1` clocks.

lpawelcz replied:
Here is the usage of this property in [GPIO driver](https://github.com/zephyrproject-rtos/zephyr/blob/main/drivers/gpio/gpio_gecko.c#L397) and in [LEUART driver usage#0](https://github.com/zephyrproject-rtos/zephyr/blob/main/drivers/serial/leuart_gecko.c#L330) + [LEUART driver usage#1](https://github.com/zephyrproject-rtos/zephyr/blob/main/drivers/serial/leuart_gecko.c#L383).
You can also see this in [UART driver](https://github.com/zephyrproject-rtos/zephyr/blob/main/drivers/serial/uart_gecko.c#L538), [SPI driver](https://github.com/zephyrproject-rtos/zephyr/blob/main/drivers/spi/spi_gecko.c#L332) and [Watchdog driver](https://github.com/zephyrproject-rtos/zephyr/blob/main/drivers/watchdog/wdt_gecko.c#L284).

It seems that we always use this property alongside the instance ID but if we start using it instead of `peripheral_id` from DTS then e.g. UART driver won't work at all so in the case of `uart_gecko.c` the driver instance ID is not the same as `perpheral_id` from DTS nodes.

gmarull replied:
As suspected, this is used because of how HAL works (it takes index 0 for gpioa, 1 for gpiob, etc.). Ref https://github.com/antmicro/hal_silabs/blob/bg22-minimal/gecko/emlib/inc/em_gpio.h#L1045. So, this is not hardware description, but DT HALization. You can create a match table in your driver between DT entries and HAL expectations (or don't use HALs :-) ).

cc: @galak

lpawelcz replied:
Yes, so I marked `peripheral-id` property in GPIO and UART bindings as not required and modified drivers to not use this property when it's not defined for given node.

gmarull commented on dts/arm/silabs/efr32bg22.dtsi:
```diff
@@ -0,0 +1,178 @@
+
+/ {
+	pinctrl: pin-controller {
+		/* Pin controller is a "virtual" device since nRF SoCs do pin
```
`nRF`?

lpawelcz replied:
Fixed, should be SiLabs.

gmarull commented on dts/arm/silabs/efr32bg22c224f512im40.dtsi:
```diff
@@ -0,0 +1,25 @@
+#include <mem.h>
+#include <silabs/efr32bg22.dtsi>
+
+/ {
+	sram0: memory@20000000 {
+		reg = <0x20000000 DT_SIZE_K(32)>;
+	};
+
+	soc {
+		compatible = "silabs,efr32bg22c224f512im40", "silabs,efr32bg22",
+						"silabs,efr32", "simple-bus";
+
+		flash-controller@40030000 {
+			flash0: flash@0 {
+				reg = <0 DT_SIZE_K(512)>;
+			};
+		};
+	};
+};
```
why not extending using nodelabels, `&flash0`, `sram0`? see GD32 for example

lpawelcz replied:
Fixed

gmarull commented on soc/arm/silabs_exx32/common/pinctrl_soc.h:
```diff
@@ -0,0 +1,92 @@
+
+#include <zephyr/devicetree.h>
+#include <zephyr/dt-bindings/pinctrl/gecko-pinctrl.h>
+#include <zephyr/types.h>
```
`<stdint.h>`?

lpawelcz replied:
Replaced include: `stdint.h` instead of `types.h`

gmarull commented on soc/arm/silabs_exx32/common/soc.c:
```diff
@@ -209,10 +255,23 @@ static int silabs_exx32_init(const struct device *arg)
 
+#ifdef CONFIG_SOC_GECKO_DEV_INIT
+#ifdef CONFIG_PM
+void enableIRQ(void)
```
this function name is misleading

lpawelcz replied:
Renamed to `initIRQ()`

gmarull commented on soc/arm/silabs_exx32/common/soc_power.c:
```diff
@@ -24,6 +69,52 @@ __weak void pm_state_set(enum pm_state state, uint8_t substate_id)
+
+	switch (state) {
+
+	case PM_STATE_RUNTIME_IDLE:
+	case PM_STATE_SUSPEND_TO_IDLE:
+	case PM_STATE_STANDBY:
```
each power state must have its own logic, just add the supported ones.

lpawelcz replied:
Fixed

gmarull commented on soc/arm/silabs_exx32/common/soc_power.c:
```diff
@@ -24,6 +69,52 @@ __weak void pm_state_set(enum pm_state state, uint8_t substate_id)
+
+	uint32_t BasePRI;
+
+	/* WorkAround: When this function is entered the Kernel has disabled
+	 * interrupts using BASEPRI register. This is incorrect as it prevents
+	 * waking up from any interrupt which priority is not 0. Work around the
+	 * issue using __set_BASEPRI(0) to remove the BASEPRI masking
+	 */
+
+	BasePRI = __get_BASEPRI();
+	__set_BASEPRI(0);
```
not sure I follow here, please explain this workaround better. It may be better to fix/adjust common code.

gmarull commented on soc/arm/silabs_exx32/common/soc_power.c:
```diff
@@ -63,3 +157,136 @@ __weak void pm_state_exit_post_ops(enum pm_state state, uint8_t substate_id)
+				pm_min_residency[i].state, ticks,
+				pm_min_residency[i].min_residency_us);
+
+			/* Busy waiting for 1 tick to flush UART buffer */
+			k_busy_wait(k_ticks_to_us_floor32(1));
+			IdleTicks = ticks - 1;
```
what's going on with UART here?

lpawelcz replied:
It looks like the UART peripheral ends in some kind of stale state if we don't shorten the sleep time. This does look like some kind of race condition, I'm looking into this.

gmarull replied:
You need to lock PM states if UART prevents from using them. Some UART drivers do this.

lpawelcz replied:
That was exactly the case here. Thank you for your suggestion, it helped me a lot!

gmarull commented on soc/arm/silabs_exx32/common/soc_power.c:
```diff
@@ -63,3 +157,136 @@ __weak void pm_state_exit_post_ops(enum pm_state state, uint8_t substate_id)
+}
+
+#ifndef CONFIG_PM_POLICY_DEFAULT
+/* CONFIG_PM_POLICY_CUSTOM allows to define application specific power policy
+ * checking whether the ticks is meeting the minimum criteria to enter
+ * into SOC specific power policy or not
+ */
+struct pm_state_info *pm_policy_next_state(uint8_t cpu, int32_t ticks)
```
please justify why a custom policy is needed.

lpawelcz replied:
I've checked the PM related code and it seems it can be refactored in a way I can get rid off the custom policy. This will require quite serious refactorization of the timer driver code. I'm working on this now and will update the PR soon.

lpawelcz replied:
It turned out that by fixing UART driver (enabling interrupt API by default and locking PM states) we can get rid of the custom policy and related code. We also don't need to modify timer driver in any way.
I've updated the PR with my modifications

gmarull reviewed (changes requested):
A few comments. PM code needs, at least, a better explanation.

lpawelcz commented:
@gmarull thank you for your review. I've already addressed most of the comments. After positive review I will squash fixes with initial set of commits from this PR.

lpawelcz committed:
drivers: gpio: gecko: stop using deprecated function

change GPIO_IntConfig to GPIO_ExtIntConfig

Signed-off-by: Paweł Czarnecki <pczarnecki@antmicro.com>

lpawelcz commented:
Hi @gmarull, I addressed all of your review comments. Please have a look at my changes and let me know if we can treat discussions in this PR as solved.

gmarull commented on boards/arm/efr32bg_sltb010a/Kconfig.defconfig:
```diff
@@ -0,0 +1,21 @@
+config CMU_LFXO_FREQ
+	default 32768
+
+config UART_GECKO
+	default y
+	depends on SERIAL
```
redundant with DT selection

lpawelcz commented on boards/arm/efr32bg_sltb010a/board.c:
```diff
@@ -0,0 +1,31 @@
+		LOG_ERR("Virtual COM Port Enable device was not found!\n");
+		return -ENODEV;
+	}
+	gpio_pin_configure_dt(&vcen_dev, GPIO_OUTPUT_HIGH);
```
Applied the change.

lpawelcz commented on boards/arm/efr32bg_sltb010a/efr32bg_sltb010a.dts:
```diff
@@ -0,0 +1,123 @@
+	aliases {
+		led0 = &led0;
+		sw0 = &button0;
+		vcomenable = &vcen;
```
The node was named poorly. The pin that is controlled by it has alternative function of wake-up trigger from the highest energy saving mode. I reorganized the node as `zephyr,user` and named the property `wake-up-gpios`.

lpawelcz commented on boards/arm/efr32bg_sltb010a/efr32bg_sltb010a.dts:
```diff
@@ -0,0 +1,123 @@
+			gpios = <&gpioa GECKO_PIN(5) GPIO_ACTIVE_LOW>;
+			label = "VCOM Enable";
+		};
+	};
```
Yes, this is a gpio input that acts as wake-up source from one of the sleep modes.

lpawelcz commented on drivers/counter/counter_gecko_stimer.c:
```diff
@@ -0,0 +1,313 @@
+
+#include <em_cmu.h>
+#include "sl_sleeptimer.h"
+#include "sl_atomic.h"
```
Changed all non-local includes to use `<>`

lpawelcz commented on drivers/counter/counter_gecko_stimer.c:
```diff
@@ -0,0 +1,313 @@
+						dev_data->top_data.ticks,
+						top_callback,
+						(void *)&dev_data->top_data,
+						0,
```
I've ran `clang-format` on all C files affected by the PR. I will leave that in a separate commit for now.

lpawelcz replied:
Formatting fixes are now integrated with commits that were changing files that had to be formatted.

lpawelcz commented on drivers/hwinfo/Kconfig:
```diff
@@ -139,7 +139,7 @@ config HWINFO_PSOC6
 	bool "GECKO hwinfo"
 	default y
-	depends on SOC_FAMILY_EXX32 && !SOC_SERIES_EFR32MG21
+	depends on SOC_FAMILY_EXX32 && !SOC_SERIES_EFR32MG21  && !SOC_SERIES_EFR32BG22
```
Changed the dependencies and added `hwinfo` node to all SiLabs SoCs and enabled it only for the supported ones.

lpawelcz commented on drivers/serial/uart_gecko.c:
```diff
@@ -11,15 +11,23 @@
+#include <zephyr/drivers/pinctrl.h>
+#else
+#include <em_gpio.h>
+#endif /* CONFIG_PINCTRL */
```
Yes, for now there is only one SiLabs SoC that supports pinctrl and it is `EFR32BG22` which is added in this PR. The rest of the SiLabs SoCs in Zephyr don't use it now. That is why we need to keep the support for both variants, at least for some time.

lpawelcz commented on drivers/serial/uart_gecko.c:
```diff
@@ -93,8 +109,30 @@ struct uart_gecko_data {
+		data->pm_policy_state_on = true;
+		pm_policy_state_lock_get(PM_STATE_STANDBY, PM_ALL_SUBSTATES);
+	}
+}
```
We can actually get rid of those wrappers. I removed them and replaced those with plain `pm_policy_state_lock_*()` calls.

mbolivar-nordic commented on boards/arm/efr32bg_sltb010a/efr32bg_sltb010a.dts:
```diff
@@ -0,0 +1,116 @@
+		};
+	};
+
+	zephyr,user {
```
NAK. The `/zephyr,user` node is reserved for application specific use; you cannot use it in board files. This needs to be removed wherever it appears outside of applications in thi sPR.


 One option for reworking this in board files is to use a node at a fixed location and use DT_PATH to get the node identifier in your C code.

gmarull replied:
note we use it now in a few boards, I guess we need to use board level compatibles then?

lpawelcz replied:
For now, I left the old `gpio-keys` compatible and renamed the node according to its function. I will prepare board level compatible for that and I will update the PR soon.

mbolivar-nordic replied:
>  I guess we need to use board level compatibles then?

yes

lpawelcz replied:
I changed this to use board level compatible so it should be OK now.

lpawelcz committed:
drivers: watchdog: silabs: include zephyr/irq.h

Add missing include

Signed-off-by: Pawel Czarnecki <pczarnecki@antmicro.com>

lpawelcz commented on drivers/pinctrl/pinctrl_gecko.c:
```diff
@@ -0,0 +1,94 @@
+ */
+
+#include <zephyr/drivers/pinctrl.h>
+#include <soc_gpio.h>
```
I deleted the `soc_gpio_configure()` function and replaced all calls to it with `GPIO_PinModeSet()`

lpawelcz commented on drivers/serial/uart_gecko.c:
```diff
@@ -176,25 +220,53 @@ static void uart_gecko_irq_tx_enable(const struct device *dev)
+
 	USART_IntEnable(config->base, mask);
+
+	irq_unlock(key);
```
I think we can remove those

lpawelcz commented:
Starting from yesterday (13.10.2022) I'd see rather strange failure here in CI, e.g.: https://github.com/zephyrproject-rtos/zephyr/actions/runs/3248323211/jobs/5329395390
It seems like it is unrelated to the PR (failure on board `mps3_an547`):
```
INFO: Error found: tests/lib/cmsis_dsp/transform/libraries.cmsis_dsp.transform.cq31.fpu on mps3_an547 (Not enough FLASH but is one of the integration platforms)
```

It also happens in other PRs: https://github.com/zephyrproject-rtos/zephyr/actions/runs/3247296293/jobs/5327088678

lpawelcz commented:
Hi @gmarull, please have a look at this PR once again. I addressed all of your comments and thanks to https://github.com/zephyrproject-rtos/zephyr/pull/51336 we've got working CI.

lpawelcz commented:
Hi @gmarull, sorry for pinging you again but this PR is ready for your review. Could you please take a look at it?

lpawelcz committed:
soc: arm: silabs: remove soc_gpio_configure wrapper

It would be better to use GPIO_PinModeSet() functions directly
in the drivers.

Signed-off-by: Pawel Czarnecki <pczarnecki@antmicro.com>

lpawelcz commented on drivers/pinctrl/pinctrl_gecko.c:
```diff
@@ -26,14 +26,16 @@ int pinctrl_configure_pins(const pinctrl_soc_pin_t *pins, uint8_t pin_cnt, uintp
 			rxpin.out = 1;
-			soc_gpio_configure(&rxpin);
+			GPIO_PinModeSet(rxpin.port, rxpin.pin, rxpin.mode,
+								rxpin.out);
```
Sure, it should be OK now

gmarull commented on soc/arm/silabs_exx32/common/soc_gpio.h:
```diff
@@ -25,12 +25,6 @@ struct soc_gpio_pin {
- * @brief Configure GPIO pin
- * @param[in] pin configuration data
- */
-void soc_gpio_configure(const struct soc_gpio_pin *pin);
```
where is `soc_gpio_pin` struct used now? I guess it can be deleted, and drivers include `<em_gpio.h>`

lpawelcz replied:
This struct is used in drivers that are compatible with other SiLabs boards that use non-pinctrl mode for GPIO configuration, see for example [UART driver](https://github.com/antmicro/zephyr/blob/efr32bg22-initial/drivers/serial/uart_gecko.c#L326-L329) or [SPI driver](https://github.com/antmicro/zephyr/blob/efr32bg22-initial/drivers/spi/spi_gecko.c#L191-L196) (other drivers are: `i2c` and `eth`). I've seen similar solutions for [atmel_sam](https://github.com/antmicro/zephyr/blob/efr32bg22-initial/soc/arm/atmel_sam/common/soc_gpio.h#L85-L94) and [cypress](https://github.com/antmicro/zephyr/blob/efr32bg22-initial/soc/arm/cypress/common/soc_gpio.h#L67-L71) SoCs. I think it is a good idea to leave this structure because it allows us to unclutter driver configuration structures from 4 different values that are specific to GPIO configuration (GPIO port, pin, mode, out register value) per each GPIO pin used in the driver (see [UART config struct](https://github.com/antmicro/zephyr/blob/efr32bg22-initial/drivers/serial/uart_gecko.c#L89-L90)).

dcpleung commented on drivers/serial/uart_gecko.c:
```diff
@@ -127,6 +129,8 @@ static void uart_gecko_poll_out(const struct device *dev, unsigned char c)
 {
 	const struct uart_gecko_config *config = dev->config;
 
+	pm_policy_state_lock_get(PM_STATE_STANDBY, PM_ALL_SUBSTATES);
```
All of these should be `#ifdef` under `CONFIG_PM`.

gmarull replied:
> All of these should be `#ifdef` under `CONFIG_PM`.

not needed, stubs are provided (see `zephyr/pm/policy.h`)

lpawelcz replied:
Yes, indeed, that is what we assumed

lpawelcz commented on drivers/serial/Kconfig.gecko:
```diff
@@ -9,6 +9,7 @@ config UART_GECKO
 	depends on DT_HAS_SILABS_GECKO_UART_ENABLED || DT_HAS_SILABS_GECKO_USART_ENABLED
 	select SERIAL_HAS_DRIVER
 	select SERIAL_SUPPORT_INTERRUPT
+	select UART_INTERRUPT_DRIVEN
```
Moved this select to soc-specific Kconfig as an extension to `CONFIG_PM` so it will be selected only for EFR32BG22 and only when PM is enabled.

lpawelcz commented:
Responded to all remaining comments and fixed `UART_INTERRUPT_DRIVEN` Kconfig selection. Please let me know if there is anything else to address here.

gmarull commented on samples/drivers/counter/alarm/src/main.c:
```diff
@@ -74,7 +78,11 @@ static void test_counter_interrupt_fn(const struct device *counter_dev,
 void main(void)
 {
-	const struct device *const counter_dev = DEVICE_DT_GET(TIMER);
+#if defined(CONFIG_COUNTER_GECKO_STIMER)
+	const struct device *counter_dev = DEVICE_DT_GET(STIMER);
+#else
+	const struct device *counter_dev = DEVICE_DT_GET(TIMER);
+#endif
```
why is this needed?

lpawelcz replied:
This was initially placed there because of warnings about naming conflicts caused by `TIMER(n)` macro defined in SiLabs HAL. This macro is actually not used so I decided to rename it (https://github.com/zephyrproject-rtos/hal_silabs/pull/19/commits/81c3c5b6b8a57a8be1ce22b58c2f9c7e76096936). Now we don't have to use the `#if` you pointed out and I was able to remove it.

lpawelcz committed:
dts: gpio: silabs: make peripheral-id optional

peripheral-id property should be eventually removed entirely.
For now set it as optional and allow skipping the usage
in GPIO driver.

Signed-off-by: Pawel Czarnecki <pczarnecki@antmicro.com>

lpawelcz committed:
dts: uart: silabs: make peripheral-id optional

peripheral-id property should be eventually removed.
For now set it as optional and allow skipping the usage
in UART driver.

Signed-off-by: Pawel Czarnecki <pczarnecki@antmicro.com>

kmeinhar commented on drivers/serial/uart_gecko.c:
```diff
@@ -9,18 +9,34 @@
+#define GET_GECKO_USART_CLOCK(id) CLOCK_USART(DT_INST_PROP(id, peripheral_id))
+#define GET_GECKO_UART_CLOCK(id) CLOCK_UART(DT_INST_PROP(id, peripheral_id))
+#else
+/* HAL expectation: default USART for UART driver is usart1 */
+#define GET_GECKO_USART_CLOCK(id) CLOCK_USART(1)
+#define GET_GECKO_UART_CLOCK(id) CLOCK_UART(id)
```
Is there a reason why USART clock is hard coded to be the UASRT1 clock?
When I am using both USART0 and USART1 the MCU will crash when initializing USART0.
I am using an EFR32PG22 and the following change resolved this issue for me:
```suggestion
#define GET_GECKO_USART_CLOCK(id) CLOCK_USART(id)
#define GET_GECKO_UART_CLOCK(id) CLOCK_UART(id)
```

lpawelcz replied:
Thank you for pointing that out. This actually strongly depends on how the USART will be used in the system. For example you can use USART0 as SPI and USART1 as UART. In such situation when instantiating the UART driver it will have `instance_id == 0` because it is the first and only UART in the system and that will cause conflicts with straightforward mapping `CLOCK_USART(id)`. You can of course switch the scheme and use USART1 as SPI but then the problem will be moved to the SPI driver. However, this particular setup is valid for Thunderboard EFR32BG22 platform in one specific configuration.
I will remove the hardcoded mapping and replace it with something more robust.

lpawelcz replied:
I changed the hardcoded values to mapping based on USART base addresses taken from defines in HAL. This should now work for all SiLabs platforms in Zephyr.

alexanderwachter reviewed (dismissed):
hwinfo looks good to me

lpawelcz committed:
dts: arm: silabs: efr32mg21: update peripheral reg addresses

HAL update affects also EFR32MG21 SoC. Because of that we need to
update the reg addresses in DTS.

Signed-off-by: Pawel Czarnecki <pczarnecki@antmicro.com>

lpawelcz commented:
Sorted out the issue with `peripheral-id` property, current solution should work on all SiLabs platforms. @gmarull please take a look into changes in https://github.com/zephyrproject-rtos/zephyr/pull/50352/commits/3d63b75d533255080b146324df8f373bbed80a38 and https://github.com/zephyrproject-rtos/zephyr/pull/50352/commits/d13c92408dc1e37cbbd36c8b75585b83388b7ee5 and let me know if this is ok for you.

gmarull commented on boards/arm/efr32bg_sltb010a/board.c:
```diff
@@ -0,0 +1,31 @@
+#include <zephyr/drivers/gpio.h>
+#include <zephyr/init.h>
+#include <zephyr/logging/log.h>
+#include <zephyr/sys/printk.h>
```
this file is not printking.

lpawelcz replied:
Include removed.

gmarull commented on boards/arm/efr32bg_sltb010a/board.c:
```diff
@@ -0,0 +1,31 @@
+		LOG_ERR("Wake-up GPIO device was not found!\n");
+		return -ENODEV;
+	}
+	gpio_pin_configure_dt(&wake_up_gpio_dev, GPIO_OUTPUT_ACTIVE);
+
+	return 0;
```
error handling ignored here.

lpawelcz replied:
Added error handling

gmarull commented on west.yml:
```diff
@@ -114,7 +114,7 @@ manifest:
       groups:
         - hal
     - name: hal_silabs
-      revision: 1ec8dd99aa4ac3e8632d2aa28a7438049bb27102
```
shouldn't hal changes come first to not break bisectability?

lpawelcz replied:
Moved HAL changes to the beginning of the PR.

gmarull commented on dts/arm/silabs/efr32mg21.dtsi:
```diff
@@ -41,9 +41,9 @@
 	};
 
 	soc {
-		msc: flash-controller@40030000 {
```
can you explain why a HAL update affects register addresses?

lpawelcz replied:
Addresses starting from `0x40004000` to `0x4C022000` are handled by Security Management Unit as secured memory regions while addresses from `0x50004000` to `0x5C022000` are non-secure.
After HAL update, register addresses had to be changed because the defines for the base addresses in HAL were renamed.
Some of the existing drivers use HAL functions that rely on defines from HAL. Previously, those defines were pointing to secured address space but after the rename those are in non-secured space by default. That caused conflicts with reg addresses in device tree and e.g. UART driver stopped working. I had to modify DTS to fix this.

lpawelcz commented on drivers/pinctrl/pinctrl_gecko.c:
```diff
@@ -0,0 +1,103 @@
+{
+#ifdef CONFIG_UART_GECKO
+	struct soc_gpio_pin rxpin = {0, 0, 0, 0};
+	struct soc_gpio_pin txpin = {0, 0, 0, 0};
```
Changed to {0}.

lpawelcz commented on dts/bindings/pinctrl/silabs,gecko-pinctrl.yaml:
```diff
@@ -0,0 +1,80 @@
+          /* group 1 ('group1' name is arbitrary) */
+          group1 {
+            /* configure P0.1 as UART_TX and P0.2 as UART_RTS */
+            psels = <GECKO_PSEL(UART_TX, PORT_A, PIN(1))>, <GECKO_PSEL(UART_RTS, PORT_A, PIN(2))>;
```
Yes, I modified this as you requested.

lpawelcz commented on soc/arm/silabs_exx32/common/soc.c:
```diff
@@ -209,10 +254,21 @@ static int silabs_exx32_init(const struct device *arg)
 
+#ifdef CONFIG_SOC_GECKO_DEV_INIT
+#ifdef CONFIG_PM
+void initIRQ(void)
```
This was removed. I will add PM code in a separate PR.

lpawelcz commented on soc/arm/silabs_exx32/common/soc.c:
```diff
@@ -209,10 +254,21 @@ static int silabs_exx32_init(const struct device *arg)
+#ifdef CONFIG_PM
+void initIRQ(void)
+{
+	IRQ_DIRECT_CONNECT(HFXO0_IRQn, 2, sl_hfxo_manager_irq_handler, 0);
```
This was removed. I will add PM code in a separate PR.

lpawelcz commented on soc/arm/silabs_exx32/common/soc_power.c:
```diff
@@ -24,6 +54,62 @@ __weak void pm_state_set(enum pm_state state, uint8_t substate_id)
+	 * offsets the priorities by 1.
+	 * This function call will effectively set the priority to 0.
+	 */
+	z_arm_irq_priority_set(RTCC_IRQn, -1, 0);
```
This was removed. I will add PM code in a separate PR.

lpawelcz commented on soc/arm/silabs_exx32/efr32bg22/Kconfig.defconfig.series:
```diff
@@ -0,0 +1,29 @@
+
+choice PM_POLICY
+	default PM_POLICY_DEFAULT
+	depends on PM
```
Removed this entry.

lpawelcz commented on soc/arm/silabs_exx32/efr32bg22/Kconfig.series:
```diff
@@ -0,0 +1,27 @@
+	select SOC_GECKO_EMU
+	select SOC_GECKO_GPIO
+	select SOC_GECKO_RTCC
+	select COUNTER_GECKO_STIMER
```
Yes, GPIO driver actually do that. I was able to remove selections:
* SOC_GECKO_EMU
* CRYPTO_ACC_GECKO_TRNG
* SOC_GECKO_GPIO
* SOC_GECKO_RTCC
* SOC_GECKO_STIMER

lpawelcz commented on soc/arm/silabs_exx32/efr32bg22/linker.ld:
```diff
@@ -0,0 +1,16 @@
+ * This is the linker script for both standard images.
+ */
+
+#include <autoconf.h>
```
Include removed.

lpawelcz commented on soc/arm/silabs_exx32/efr32bg22/soc.h:
```diff
@@ -0,0 +1,26 @@
+#ifndef EFR32BG22_SOC_H_
+#define EFR32BG22_SOC_H_
+
+#include <zephyr/sys/util.h>
```
It is unnecessary, removed it.

lpawelcz commented on soc/arm/silabs_exx32/efr32bg22/soc.h:
```diff
@@ -0,0 +1,26 @@
+
+#ifndef _ASMLANGUAGE
+
+#include "../common/soc_gpio.h"
```
It is unnecessary, removed it.

lpawelcz commented:
It seems like [failures](https://github.com/zephyrproject-rtos/zephyr/pull/50352/checks?check_run_id=9478691624) in CI are unrelated to my changes and are caused by warnings from building `tests/drivers/counter/counter_basic_api` on STM platforms:
* stm32f103_mini
* nucleo_l4a6zg
* nucleo_f103rb

Created https://github.com/zephyrproject-rtos/zephyr/issues/52235 for that.

gmarull commented on dts/arm/silabs/efr32bg22c224f512im40.dtsi:
```diff
@@ -0,0 +1,24 @@
+
+/ {
+	soc {
+		compatible = "silabs,efr32bg22c224f512im40", "silabs,efr32bg22",
+						"silabs,efr32", "simple-bus";
```
nit: fix alignment here

lpawelcz replied:
Fixed

gmarull commented on soc/arm/silabs_exx32/common/soc.c:
```diff
@@ -189,6 +200,18 @@ static int silabs_exx32_init(const struct device *arg)
+	sl_device_init_dpll();
+	sl_device_init_clocks();
+	sl_device_init_emu();
+	sl_board_init();
```
I've just checked what sl_board_init() does, and we can't definitely accept something like this in Zephyr. https://github.com/zephyrproject-rtos/hal_silabs/pull/19/files#diff-732e868df2ebaa0cb659a0f52f366c33d4b9f9bbb44416e351732fc24df87c40R45

Zephyr already has a mechanism to do these things (boards), and this clearly deviates from what IMO is acceptable. How is going to be generic for all Zephyr boards if put under soc?

Unfortunately, we have historically pulled HAL code without control/audit to make "progress", and this slows progress on core features later (e.g. pinctrl, clock control, device model, etc.).

lpawelcz replied:
@gmarull from what I know, the purpose of the whole init process enclosed under `#ifdef CONFIG_SOC_GECKO_DEV_INIT` is meant to represent the init process for the board as it is done in Simplicity Studio. I understand that it is not the way it should be done in Zephyr. Initially it was done this way to speed up the develepment. Would it be ok for now if I move this sequence to board-specific init?

lpawelcz replied:
`sl_board_preinit` was completely removed. Instead of this I placed GPIO clock enable in the common initialization part of the GPIO driver. I also had to set the GPIO driver initialization stage to `EARLY` - UART driver depends on correct GPIO clock configuration and UART driver is initialized at `PRE_KERNEL_1`.
`sl_board_init` and `sl_device_init_clocks` were moved to board init function in `board.c` because those functions are board specific.

gmarull commented on soc/arm/silabs_exx32/common/soc.c:
```diff
@@ -189,6 +200,18 @@ static int silabs_exx32_init(const struct device *arg)
 
+#ifdef CONFIG_SOC_GECKO_DEV_INIT
+	CHIP_Init();
+	sl_board_preinit();
```
this call seems to be initializing GPIO clocks. This must be done by the GPIO driver.

lpawelcz replied:
Moved initializing GPIO clocks to GPIO driver.

gmarull commented on soc/arm/silabs_exx32/efr32bg22/Kconfig.defconfig.series:
```diff
@@ -0,0 +1,24 @@
+	default 60
+
+config PM
+	default n
```
redundant

gmarull commented on soc/arm/silabs_exx32/efr32bg22/Kconfig.defconfig.series:
```diff
@@ -18,6 +18,7 @@ config NUM_IRQS
 config PM
 	default n
 	select COUNTER
+	select UART_INTERRUPT_DRIVEN
```
since PM has been dropped, you can likely drop this for now

lpawelcz replied:
Removed this

gmarull commented on drivers/hwinfo/Kconfig:
```diff
@@ -139,7 +139,7 @@ config HWINFO_PSOC6
 	bool "GECKO hwinfo"
 	default y
-	depends on SOC_FAMILY_EXX32 && !SOC_SERIES_EFR32MG21  && !SOC_SERIES_EFR32BG22
+	depends on SOC_FAMILY_EXX32 && DT_HAS_SILABS_GECKO_HWINFO_ENABLED
```
SOC dependency is redundant now

lpawelcz replied:
removed `SOC_FAMILY_EXX32` dependency

gmarull commented on drivers/gpio/gpio_gecko.c:
```diff
@@ -19,6 +19,22 @@
+ * 1. Address space of the first GPIO port starts at GPIO peripheral
+ *    base address which is taken from GPIO_BASE HAL macro
+ */
+#define GET_GECKO_GPIO_INDEX(id) (DT_INST_REG_ADDR(id) - GPIO_BASE) \
```
`GPIO_BASE` -> `DT_REG_ADDR(DT_NODELABEL(gpioa))`, no HAL needed for addresses.

lpawelcz replied:
Changed `GPIO_BASE` to `DT_REG_ADDR(DT_NODELABEL(gpioa))`

gmarull commented on drivers/serial/uart_gecko.c:
```diff
@@ -20,10 +20,54 @@
 #define UART_PREFIX cmuClock_UART
 #define CLOCK_USART(id) _CONCAT(USART_PREFIX, id)
 #define CLOCK_UART(id) _CONCAT(UART_PREFIX, id)
+#define GET_GECKO_USART_CLOCK(id) CLOCK_USART(DT_INST_PROP(id, peripheral_id))
+#define GET_GECKO_UART_CLOCK(id) CLOCK_UART(DT_INST_PROP(id, peripheral_id))
+#else
+#if (USART_COUNT <= 2)
+#define CLOCK_USART(ref)	(((ref) == USART0) ? cmuClock_USART0 \
+			       : ((ref) == USART1) ? cmuClock_USART1 \
+			       : -1)
+#elif (USART_COUNT == 3)
+#define CLOCK_USART(ref)	(((ref) == USART0) ? cmuClock_USART0 \
+			       : ((ref) == USART1) ? cmuClock_USART1 \
+			       : ((ref) == USART2) ? cmuClock_USART2 \
+			       : -1)
+#elif (USART_COUNT == 4)
+#define CLOCK_USART(ref)	(((ref) == USART0) ? cmuClock_USART0 \
```
non-blocking, but this clearly shows clock information needs to be captured properly in DT using `clocks` prop.

lpawelcz replied:
Right, I will change that in the follow-up PR if that's OK for you.

lpawelcz commented on drivers/serial/uart_gecko.c:
```diff
@@ -126,6 +128,8 @@ static void uart_gecko_poll_out(const struct device *dev, unsigned char c)
 {
 	const struct uart_gecko_config *config = dev->config;
 
+	pm_policy_state_lock_get(PM_STATE_STANDBY, PM_ALL_SUBSTATES);
```
Removed `2c83dcba` and `51cf90a4`

lpawelcz commented on soc/arm/silabs_exx32/efr32bg22/Kconfig.defconfig.efr32bg22:
```diff
@@ -0,0 +1,20 @@
+
+config SPI_GECKO
+	default n
+	depends on SPI
```
Removed unnecessary dependencies. In fact, those configs already depend on `GPIO`, `I2C`, etc. through `if GPIO`, `if I2C` statements in main driver Kconfigs.

lpawelcz commented on dts/arm/silabs/efr32bg22.dtsi:
```diff
@@ -0,0 +1,170 @@
+			device_type = "cpu";
+			compatible = "arm,cortex-m33";
+			reg = <0>;
+			cpu-power-states = < &standby >;
```
Fixed

lpawelcz commented on dts/arm/silabs/efr32bg22-pinctrl.dtsi:
```diff
@@ -0,0 +1,29 @@
+			/* configure PE.11 as UART_RX and PE.10 as UART_TX */
+			psels = <GECKO_PSEL(UART_TX, E, 10)>,
+				<GECKO_PSEL(UART_RX, E, 11)>,
+				<GECKO_LOC(UART_LOC, GECKO_LOCATION(0))>;
```
Done

lpawelcz commented on samples/drivers/counter/alarm/src/main.c:
```diff
@@ -33,13 +33,17 @@ struct counter_alarm_cfg alarm_cfg;
 				      void *user_data)
 {
-	struct counter_alarm_cfg *config = user_data;
+	struct counter_alarm_cfg *const config = user_data;
```
This was accidentally added while doing cleanup, fixed this.

lpawelcz commented on samples/drivers/counter/alarm/src/main.c:
```diff
@@ -74,7 +78,7 @@ static void test_counter_interrupt_fn(const struct device *counter_dev,
 void main(void)
 {
-	const struct device *const counter_dev = DEVICE_DT_GET(TIMER);
+	const struct device *counter_dev = DEVICE_DT_GET(TIMER);
```
This was accidentally removed while doing cleanup, fixed this.

lpawelcz committed:
drivers: gpio: gecko: initialize driver earlier

Common GPIO driver needs to be initialized before e.g UART driver
to ensure correct GPIO clock configuration when configuring UART pinout

Signed-off-by: Pawel Czarnecki <pczarnecki@antmicro.com>

lpawelcz committed:
drivers: gpio: gecko: enable GPIO clock

Enable GPIO clock so that other drivers could
configure their I/O pins

Signed-off-by: Pawel Czarnecki <pczarnecki@antmicro.com>

lpawelcz commented:
@gmarull really appreciate all the good feedback!
CI failures seem to have been fixed now in https://github.com/zephyrproject-rtos/zephyr/pull/52279. I rebased and now all tests have passed. Could you please take another look at this PR and assuming everything looks OK, approve it?
This is the first of a series of PRs which we have prepared / are preparing which target Silicon Labs platforms. Having this initial PR in place would allow us to subsequently improve the support with fixes, drivers and other platforms.

gmarull commented on drivers/gpio/gpio_gecko.c:
```diff
@@ -366,7 +366,7 @@ DEVICE_DT_DEFINE(DT_INST(0, silabs_gecko_gpio),
 		    NULL,
 		    &gpio_gecko_common_data, &gpio_gecko_common_config,
-		    POST_KERNEL, CONFIG_GPIO_GECKO_COMMON_INIT_PRIORITY,
+		    EARLY, CONFIG_GPIO_GECKO_COMMON_INIT_PRIORITY,
```
EARLY level is meant for code that needs to run at very early stages, just after C domain is ready. Adjust priorities so that GPIO comes before UART instead. Or, if UART driver needs GPIO clocks, make sure it turns them on (better probably since you have an implicit dependency that is not captured anywhere).

lpawelcz replied:
Changed the initialization stage from `EARLY` to `PRE_KERNEL_1` (the same as UART driver). Default INIT_PRIORITY for common part of the GPIO driver has priority `39` while for the UART driver it is `50` (priorities remain unchanged).

gmarull commented on boards/arm/efr32bg_sltb010a/board.c:
```diff
@@ -0,0 +1,43 @@
+
+#ifdef CONFIG_SOC_GECKO_DEV_INIT
+	sl_device_init_clocks();
+	sl_board_init();
```
definitely a -1 to run code like this https://github.com/antmicro/hal_silabs/blob/bg22-minimal/gecko/board/src/sl_board_init.c. This is bypassing Zephyr mechanisms, including Devicetree as a source of hardware description.

lpawelcz replied:
We can actually remove this function call entirely. GPIO clock is enabled by GPIO driver and the rest of the function (peripheral init) is not needed.

lpawelcz commented:
@gmarull thank you for your comments. I addressed those and rebased the code. I also updated pinctrl bindings and added entry in `CODEOWNERS` file because compliance checks were failing in CI. Please let me know if those changes are OK.

alexanderwachter commented on drivers/hwinfo/Kconfig:
```diff
@@ -139,7 +139,7 @@ config HWINFO_PSOC6
 	bool "GECKO hwinfo"
 	default y
-	depends on SOC_FAMILY_EXX32 && !SOC_SERIES_EFR32MG21
+	depends on DT_HAS_SILABS_GECKO_HWINFO_ENABLED
```
I'm not sure if we should use the device-tree for this. hwinfo has no device instances, no register address, ...
@mbolivar-nordic do you know of any driver where we use the device-tree to only enable it?

lpawelcz replied:
Hi @alexanderwachter, there is another hwinfo driver enabled with devicetree [here](https://github.com/zephyrproject-rtos/zephyr/blob/main/drivers/hwinfo/Kconfig#L70). Also some time earlier in this PR @gmarull suggested to use [DT_HAS_* defines](https://github.com/zephyrproject-rtos/zephyr/pull/50352/commits/cbe8bd62ed4e77b563c289453cffbcf536025363#r991082288). However, if that approach is not acceptable we can change it back to:
```
depends on SOC_FAMILY_EXX32 && !SOC_SERIES_EFR32MG21 && !SOC_SERIES_EFR32BG22
```

alexanderwachter replied:
I just want feedback from the device-tree experts. I'm ok with both.

gmarull replied:
Looks like driver is using RMU (Reset Management Unit?), but all hw info is taken directly from HAL, bypassing DT. I guess we need to use SOC_SERIES* in this case, and use a non-DT device as well. Yet another example where HALs do more harm than good for the consistency of the project.

lpawelcz replied:
Removed commits https://github.com/zephyrproject-rtos/zephyr/pull/50352/commits/33980eade4ed98bdc9ea3263a5bc162d9460769a and https://github.com/zephyrproject-rtos/zephyr/pull/50352/commits/93075ea6a47791e3c5d20ac7ebf848e66ad518cc so that we don't use devicetree for hwinfo driver.

gmarull commented on drivers/hwinfo/hwinfo_gecko.c:
```diff
@@ -37,75 +37,85 @@ int z_impl_hwinfo_get_reset_cause(uint32_t *cause)
 	uint32_t flags = 0;
 	uint32_t rmu_flags = RMU_ResetCauseGet();
 
+#if IS_ENABLED(RMU_RSTCAUSE_PORST)
```
shouldn't these be `#ifdef`?

lpawelcz replied:
Switched those back to `#ifdef`.

lpawelcz commented:
@gmarull, @alexanderwachter I addressed your latest comments. Additionally, I had to add a new header to board directory which is necessary for hfxo oscillator initialization after making changes to [HAL PR](https://github.com/zephyrproject-rtos/hal_silabs/pull/19).
Please let me know if changes in this PR look OK for you now.

gmarull commented on boards/arm/efr32bg_sltb010a/board.c:
```diff
@@ -0,0 +1,64 @@
+	return 0;
+}
+
+#ifdef CONFIG_SOC_GECKO_DEV_INIT
+int efr32bg_sltb010a_init_clocks(void)
+{
+	CMU_ClockSelectSet(cmuClock_SYSCLK, cmuSelect_HFRCODPLL);
+#if defined(_CMU_EM01GRPACLKCTRL_MASK)
+	CMU_ClockSelectSet(cmuClock_EM01GRPACLK, cmuSelect_HFRCODPLL);
+#endif
+#if defined(_CMU_EM01GRPBCLKCTRL_MASK)
+	CMU_ClockSelectSet(cmuClock_EM01GRPBCLK, cmuSelect_HFRCODPLL);
+#endif
+	CMU_ClockSelectSet(cmuClock_EM23GRPACLK, cmuSelect_LFRCO);
+	CMU_ClockSelectSet(cmuClock_EM4GRPACLK, cmuSelect_LFRCO);
+#if defined(RTCC_PRESENT)
+	CMU_ClockSelectSet(cmuClock_RTCC, cmuSelect_LFRCO);
+#endif
+	CMU_ClockSelectSet(cmuClock_WDOG0, cmuSelect_LFRCO);
+
+	return 0;
+}
+#endif
```
while clock configuration is board dependent, we should not be handling this on a board basis IMO. Boards should define the clock configuration in Devicetree, and get it applied using generic code (see e.g. STM32). Anyway, clock control in Zephyr is already a mess, so I guess we can make an exception here as soon as there is a compromise to fix this later.

fkokosinski replied:
It looks like other exx32-based boards (e.g. efm32gg_stk3701a) have their clocks configured by the board init code as well. Let's land this PR first, and continue the clk cleanup in the following PRs, as we will need to modify more than just this board (efr32bg_sltb010a).

gmarull commented on soc/arm/silabs_exx32/Kconfig:
```diff
@@ -134,6 +134,18 @@ choice SOC_GECKO_EMU_DCDC_MODE
 		bool "Bypass"
 endchoice
 
+config SOC_GECKO_DEV_INIT
+	bool
+	help
+	  Device initialization as per the sequence in simplicity studio.
+	  This is currently supported by efr32bg22 series.
```
this option is not clear, it seems to translate to clock configuration. It should be clarified, and implications detailed.

fkokosinski replied:
I've changed the description of this config option.

gmarull commented on boards/arm/efr32bg_sltb010a/board.c:
```diff
@@ -0,0 +1,64 @@
+}
+
+#ifdef CONFIG_SOC_GECKO_DEV_INIT
+int efr32bg_sltb010a_init_clocks(void)
```
static missing here

fkokosinski replied:
Done.

alexanderwachter reviewed (dismissed):
hwinfo LGTM

fkokosinski commented:
Hi @gmarull, PTAL and let me know if you're OK now with the changes.

gmarull reviewed (dismissed):
I think it's mature enough to merge. However, clock control should be implemented properly soon.

fkokosinski commented:
@alexanderwachter could you PTAL? I had to dismiss your ACK to address other comments. Thanks!

fkokosinski commented:
Hi @gmarull, @cfriedt and @alexanderwachter. I had to rebase this PR due to a conflict in `samples/drivers/counter/alarm/src/main.c`. PTAL at this PR. Thanks!

gmarull commented:
> Hi @gmarull, @cfriedt and @alexanderwachter. I had to rebase this PR due to a conflict in `samples/drivers/counter/alarm/src/main.c`. PTAL at this PR. Thanks!

please merge HAL PR and update sha as well.

stephanosio commented:
https://github.com/zephyrproject-rtos/hal_silabs/pull/19 has been merged.

lpawelcz commented:
Hi @gmarull, @cfriedt, @alexanderwachter. HAL PR has been merged. I rebased the PR and updated HAL sha in `west.yml`. Please take another look at this PR. Thank you!

cfriedt commented on dts/bindings/gpio/silabs,gecko-gpio-port.yaml:
```diff
@@ -10,7 +10,7 @@ properties:
     peripheral-id:
       type: int
-      required: true
+      required: false
```
This is redundant - should have been caught by CI ..

cfriedt commented on dts/bindings/serial/silabs,gecko-usart.yaml:
```diff
@@ -13,20 +13,18 @@ properties:
     peripheral-id:
       type: int
-      required: true
+      required: false
```
This is redundant as well.

cfriedt reviewed (approved):
Just pointed out some `required: false` instances in Device Tree YAML, but not blocking. If you need to go through another rev then fix, otherwise, fix later.

stephanosio commented:
@lpawelcz This seems to have broken multiple networking tests: https://github.com/zephyrproject-rtos/zephyr/pull/52852/checks?check_run_id=10231260430. Can you provide a fix?

cfriedt commented:
> 

@stephanosio  I'll take a look if nobody else is working on it atm

lpawelcz commented:
@stephanosio @cfriedt I have the fix already. Give me a second and I will open a PR
